ads/google/core-shared.js

```diff
@@ -0,0 +1,482 @@
+ *   layoutType: (string|undefined),
+ * }} params
+ * @return {?{
+ *   numberOfRows: (!Array<number>|undefined),
+ *   numberOfColumns: (!Array<number>|undefined),
+ *   layoutTypes: (!Array<!LayoutType>|undefined),
+ * }} parsed params or null if they were invalid.
+ */
+function validateAndParsePubControlParams(params) {
+  // Verify that either all three parameters passed or none.
+  let numberOfPubControlParams = 0;
+  if (params.layoutType) {
+    numberOfPubControlParams++;
+  }
+  if (params.numberOfColumns) {
+    numberOfPubControlParams++;
+  }
+  if (params.numberOfRows) {
+    numberOfPubControlParams++;
+  }
+  if (numberOfPubControlParams < 3) {
+    reportError(`Tags ${ExternalCorePubVars.UI_TYPE}, ${
+      ExternalCorePubVars.COLUMNS_NUM} and ${
+      ExternalCorePubVars.ROWS_NUM} should be set together.`);
+    return null;
+  }
+
+  const /** !Array<!LayoutType> */ layoutTypes = params.layoutType.split(',');
+  const /** !Array<string> */ numberOfRows = params.numberOfRows.split(',');
+  const /** !Array<string> */ numberOfColumns =
+    params.numberOfColumns.split(',');
+
+  // Check all params have same length.
+  if (layoutTypes.length !== numberOfRows.length ||
+    layoutTypes.length !== numberOfColumns.length) {
+    reportError(`Lengths of parameters ${ExternalCorePubVars.UI_TYPE}, ${
+      ExternalCorePubVars.COLUMNS_NUM} and ${
+      ExternalCorePubVars.ROWS_NUM} must match. Example: ${
+      PUB_CONTROL_EXAMPLE}`);
+    return null;
+  }
+
+  if (layoutTypes.length > 2) {
+    reportError(
+        `The parameter length of attribute ${ExternalCorePubVars.UI_TYPE}, ${
+          ExternalCorePubVars.COLUMNS_NUM} and ${
+          ExternalCorePubVars
+              .ROWS_NUM} is too long. At most 2 parameters for each ` +
+      'attribute are needed: one for mobile and one for desktop, while ' +
+      `you are providing ${layoutTypes.length} parameters. Example: ${
+        PUB_CONTROL_EXAMPLE}.`);
+    return null;
+  }
+
+  const /** !Array<number> */ numberOfRowsAsNumbers = [];
+  const /** !Array<number> */ numberOfColumnsAsNumbers = [];
+  for (let i = 0; i < layoutTypes.length; i++) {
+    const row = Number(numberOfRows[i]);
+    if (isNaN(row) || row === 0) {
+      reportError(`Wrong value '${numberOfRows[i]}' for ${
+        ExternalCorePubVars.ROWS_NUM}.`);
+      return null;
+    }
+    numberOfRowsAsNumbers.push(row);
+    const col = Number(numberOfColumns[i]);
+    if (isNaN(col) || col === 0) {
+      reportError(`Wrong value '${numberOfColumns[i]}' for ${
+        ExternalCorePubVars.COLUMNS_NUM}.`);
+      return null;
+    }
+    numberOfColumnsAsNumbers.push(col);
+  }
+  return {
+    numberOfRows: numberOfRowsAsNumbers,
+    numberOfColumns: numberOfColumnsAsNumbers,
+    layoutTypes,
+    paramsInvalid: false,
+  };
+}
+
+/**
+ * @param {number} availableWidth
+ * @param {boolean} isMobile
+ * @return {!CoReConfig}
+ */
+export function getAutoConfig(availableWidth, isMobile) {
```

Comment:
please organize the file so that public function is before local function.

Reply:
Done.